In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
df = pd.read_csv('train.csv')

In [2]:
df.head()

1stFlrSF  2ndFlrSF  3SsnPorch  BedroomAbvGr  BsmtFinSF1  BsmtFinSF2  \
0       856       854          0             3         706           0   
1      1262         0          0             3         978           0   
2       920       866          0             3         486           0   
3       961       756          0             3         216           0   
4      1145      1053          0             4         655           0   

   BsmtFullBath  BsmtHalfBath  BsmtUnfSF  EnclosedPorch    ...      \
0             1             0        150              0    ...       
1             0             1        284              0    ...       
2             1             0        434              0    ...       
3             1             0        540            272    ...       
4             1             0        490              0    ...       

   SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0               0               0             0             0            1   
1               0               0             0             0            1   
2               0               0             0             0            1   
3               0               0             0             0            1   
4               0               0             0             0            1   

   Street_Grvl  Street_Pave  Utilities_AllPub  Utilities_NoSeWa  SalePrice  
0            0            1                 1                 0     208500  
1            0            1                 1                 0     181500  
2            0            1                 1                 0     223500  
3            0            1                 1                 0     140000  
4            0            1                 1                 0     250000  

[5 rows x 271 columns]

In [3]:
df.isnull().sum()

1stFlrSF                 0
2ndFlrSF                 0
3SsnPorch                0
BedroomAbvGr             0
BsmtFinSF1               0
BsmtFinSF2               0
BsmtFullBath             0
BsmtHalfBath             0
BsmtUnfSF                0
EnclosedPorch            0
Fireplaces               0
FullBath                 0
GarageArea               0
GarageCars               0
GarageYrBlt              0
GrLivArea                0
HalfBath                 0
KitchenAbvGr             0
LotArea                  0
LotFrontage              0
LowQualFinSF             0
MSSubClass               0
MasVnrArea               0
MiscVal                  0
MoSold                   0
OpenPorchSF              0
OverallCond              0
OverallQual              0
PoolArea                 0
ScreenPorch              0
                        ..
RoofMatl_Roll            0
RoofMatl_Tar&Grv         0
RoofMatl_WdShake         0
RoofMatl_WdShngl         0
RoofStyle_Flat           0
RoofStyle_Gable          0
R

In [30]:
X = df.iloc[:, :270]

In [31]:
X.head()

1stFlrSF  2ndFlrSF  3SsnPorch  BedroomAbvGr  BsmtFinSF1  BsmtFinSF2  \
0       856       854          0             3         706           0   
1      1262         0          0             3         978           0   
2       920       866          0             3         486           0   
3       961       756          0             3         216           0   
4      1145      1053          0             4         655           0   

   BsmtFullBath  BsmtHalfBath  BsmtUnfSF  EnclosedPorch        ...         \
0             1             0        150              0        ...          
1             0             1        284              0        ...          
2             1             0        434              0        ...          
3             1             0        540            272        ...          
4             1             0        490              0        ...          

   SaleType_ConLD  SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  \
0               0               0               0             0             0   
1               0               0               0             0             0   
2               0               0               0             0             0   
3               0               0               0             0             0   
4               0               0               0             0             0   

   SaleType_WD  Street_Grvl  Street_Pave  Utilities_AllPub  Utilities_NoSeWa  
0            1            0            1                 1                 0  
1            1            0            1                 1                 0  
2            1            0            1                 1                 0  
3            1            0            1                 1                 0  
4            1            0            1                 1                 0  

[5 rows x 270 columns]

In [32]:
y = df.loc[:, ['SalePrice']].values

In [33]:
X[:10]

1stFlrSF  2ndFlrSF  3SsnPorch  BedroomAbvGr  BsmtFinSF1  BsmtFinSF2  \
0       856       854          0             3         706           0   
1      1262         0          0             3         978           0   
2       920       866          0             3         486           0   
3       961       756          0             3         216           0   
4      1145      1053          0             4         655           0   
5       796       566        320             1         732           0   
6      1694         0          0             3        1369           0   
7      1107       983          0             3         859          32   
8      1022       752          0             2           0           0   
9      1077         0          0             2         851           0   

   BsmtFullBath  BsmtHalfBath  BsmtUnfSF  EnclosedPorch        ...         \
0             1             0        150              0        ...          
1             0             1        284              0        ...          
2             1             0        434              0        ...          
3             1             0        540            272        ...          
4             1             0        490              0        ...          
5             1             0         64              0        ...          
6             1             0        317              0        ...          
7             1             0        216            228        ...          
8             0             0        952            205        ...          
9             1             0        140              0        ...          

   SaleType_ConLD  SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  \
0               0               0               0             0             0   
1               0               0               0             0             0   
2               0               0               0             0             0   
3               0               0               0             0             0   
4               0               0               0             0             0   
5               0               0               0             0             0   
6               0               0               0             0             0   
7               0               0               0             0             0   
8               0               0               0             0             0   
9               0               0               0             0             0   

   SaleType_WD  Street_Grvl  Street_Pave  Utilities_AllPub  Utilities_NoSeWa  
0            1            0            1                 1                 0  
1            1            0            1                 1                 0  
2            1            0            1                 1                 0  
3            1            0            1                 1                 0  
4            1            0            1                 1                 0  
5            1            0            1                 1                 0  
6            1            0            1                 1                 0  
7            1            0            1                 1                 0  
8            1            0            1                 1                 0  
9            1            0            1                 1                 0  

[10 rows x 270 columns]

In [34]:
#データの標準化
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X) 
X_std = scaler.fit_transform(X) 

In [35]:
#主成分分析を実行
from sklearn.decomposition import PCA
pca = PCA(n_components=23)
X_pca = pca.fit_transform(X_std)

In [36]:
print(np.cumsum(pca.explained_variance_ratio_))

[ 0.06629801  0.09628442  0.12072863  0.1404985   0.15835528  0.17510829
  0.19015558  0.20290643  0.21545799  0.22732941  0.2388806   0.2498831
  0.26055612  0.27102801  0.28123235  0.29126617  0.30102195  0.31037414
  0.31956452  0.3285519   0.3374594   0.34609911  0.35464323  0.36293636
  0.37109089  0.37919263  0.38723974  0.39515672  0.40295742  0.41075142
  0.41842993  0.42597421  0.4332771   0.44051566  0.44769917  0.45469201
  0.46163242  0.46848003  0.47521941  0.4818877   0.4884673   0.49497629
  0.50137698  0.50763383  0.51386775  0.52006447  0.52613616  0.53212107
  0.53799715  0.54381575  0.54957435  0.55521367  0.56082768  0.5663675
  0.57186154  0.57728975  0.58270095  0.58801598  0.59326127  0.59846862
  0.60360764  0.60870104  0.61377004  0.61873807  0.62367879  0.62852956
  0.63333095  0.63808971  0.64282885  0.64755075  0.65219972  0.65682704
  0.66142031  0.66595284  0.67044235  0.67491359  0.67933238  0.68373389
  0.68802122  0.69229696  0.69654759  0.7007903   0.7

In [37]:
#上の中で、どこで80を超えるかチェック（20:80の法則より）
a = np.cumsum(pca.explained_variance_ratio_)

i = 0
while i < len(a):
    if a[i] < 0.8:
        i +=1
    else:
        print(i)
        break

108


In [38]:
#上のコマンドにより初めて80を超えるのは108番目だと分かったので、そこまでのデータを取り出す
X_pca = X_pca[:, :108]

In [39]:
X_pca

array([[ 4.33375061,  1.61816149, -2.16261454, ...,  0.06325218,
        -0.1210312 , -0.651761  ],
       [-0.18320543, -3.13609898, -0.64148051, ..., -1.73236047,
         0.46485707,  0.87719392],
       [ 4.84985709,  1.11805091, -1.52702288, ..., -0.30260031,
         0.45540819, -0.3809233 ],
       ..., 
       [ 0.92609846,  1.67334474,  2.27556039, ...,  0.50819825,
         0.49010655, -2.01482057],
       [-3.15680908, -3.59189763, -0.76433473, ...,  0.83259552,
         0.79400105, -1.00700547],
       [-1.30299528, -3.85578254, -1.30986787, ...,  0.70586006,
        -0.12546216,  2.22927399]])

In [40]:
from sklearn.model_selection import train_test_split
X_pca_train, X_pca_test, y_train, y_test = train_test_split(X_pca, y, test_size = 0.3, random_state = 0)

In [41]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [42]:
lr.fit(X_pca_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [44]:
def adjusted(score, n_sample, n_features):
    adjusted_score = 1 - (1 - score) * ((n_sample - 1) / (n_sample - n_features - 1))
    return adjusted_score

In [46]:
print('train: %3f' % adjusted(lr.score(X_pca_train, y_train), len(y_train), 2))
print('test : %3f' % adjusted(lr.score(X_pca_test, y_test), len(y_test), 2))

train: 0.912483
test : 0.540303


In [69]:
#ここで過学習を起こしているので、正則化！

In [188]:
from sklearn.linear_model import Lasso
model_lasso = Lasso(alpha = 2903) #alphaの数字の大きさで強さを指定
model_lasso.fit(X_pca_train,y_train) 

Lasso(alpha=2903, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [189]:
model_lasso.score(X_pca_test, y_test)

0.73962992509085412

# plt.scatter(X_pca_train, y_train)                                              
# plt.plot(X_pca_train_plot, y_train_plot, color='gray')                        
# plt.plot(X_pca_train_plot, y_train_plot_pol, color='green')                   
# plt.plot(X_pca_train_plot, model_lasso.predict(X_pca_train_plot), color='red') 
# plt.ylim([-1.6, 1.6])  
# plt.show()         

#今回の場合、説明変数が多すぎるため、グラフでは扱えない
#95でやってるように、数値で判断できるので、ただ単にグラフで検証する意味ない